edx_exams/apps/core/api.py

```diff
@@ -142,3 +144,58 @@ def get_exam_attempt_time_remaining(exam_attempt, now=None):
         secs_to_end = (end_time-now).total_seconds()
 
     return secs_to_end
+
+
+def create_exam_attempt(exam_id, user_id):
+    """
+    Creates an exam attempt for user_id against exam_id. There should only
+    be one exam_attempt per user per exam.
+    """
+
+    exam_obj = Exam.get_exam_by_id(exam_id)
+
+    if exam_obj is None:
+        err_msg = (
+            f'Exam with exam_id={exam_id} does not exist.'
+        )
+        raise ExamDoesNotExist(err_msg)
+
+    log.info(
+        ('Creating exam attempt for exam_id=%(exam_id)s for user_id=%(user_id)s '
+         'in course_id=%(course_id)s'),
+        {
+            'exam_id': exam_id,
+            'user_id': user_id,
+            'course_id': exam_obj.course_id,
+        }
+    )
+
+    practice_exam_types = [PracticeExamType, OnboardingExamType]
+
+    # if exam is past the due date, and it is a non-practice exam, raise error
+    if get_exam_type(exam_obj.exam_type) not in practice_exam_types and timezone.now() > exam_obj.due_date:
+        err_msg = (
+            f'user_id={user_id} trying to create exam attempt for past due non-practice exam '
+            f"exam_id={exam_id} in course_id={exam_obj.course_id}. Do not register an exam attempt!"
+        )
+        raise ExamAttemptOnPastDueExam(err_msg)
+
+    # create exam attempt
+    attempt = ExamAttempt.objects.create(
+        exam_id=exam_id,
+        user_id=user_id,
+        status=ExamAttemptStatus.created,
+        attempt_number=1,
```

Comment:
this needs a non-null value, but I'm not quite sure what the value should be. This is an int field, and it's needed by the LTI code, but I'm not sure if it should be auto-incrementing or if we should create a random int value.

Reply:
I'll quote the relevant [portion](http://www.imsglobal.org/spec/proctoring/v1p0#h.m0wu6zl84ted) of the proctoring specification below.

> The https://purl.imsglobal.org/spec/lti-ap/claim/attempt_number claim MUST specify the candidate's attempt number.  Used in conjunction with the resource link claim this information can be used by the Proctoring Tool to determine if the candidate is starting a new attempt at the assessment or is returning to resume a previous attempt.  The value SHOULD be simple integer with the first attempt having attempt number 1.
> 
> This specification imposes no limits on the number of times a Candidate can initiate the proctoring process for a given attempt at an assessment.  A Proctoring Tool MAY impose a limit of one successful launch per assessment (resource_link) per user.  Multiple attempts at the same assessment SHOULD NOT be assumed without prior arrangement with the Proctoring Tool.
>
> Example 4.2.3.6.1: the Candidate initiates the exam launch workflow and is transferred to the Proctoring Tool.  Once connected with the [proctor](http://www.imsglobal.org/spec/proctoring/v1p0#dfn-proctor) they are asked to present their government issued ID document but are unable to provide anything suitable.  They return the following day and initiate the exam launch workflow a second time.  The Assessment Platform sends the sameattempt number in both Start Proctoring messages.
> 
> Example 4.2.3.6.2: the Candidate initiates the assessment launch workflow for their first attempt at their certification exam.  The Assessment Platform sets the attempt number to 1 in the Start Proctoring message.  The [proctor](http://www.imsglobal.org/spec/proctoring/v1p0#dfn-proctor) is satisfied that the candidate's environment is secure and the candidate is allowed to start the assessment.  The candidate fails the exam and returns one month later.  When they initiate the launch workflow again the Assessment Platform sends attempt number 2.

Based on this, I think the value should be an auto-incrementing integer starting at 1. The specification doesn't seem to outright require that it be auto-incrementing, but it does suggest it.

Does that help?

Reply:
I think it does, but are they saying that the attempt_number is unique to each user? I.e. this should just default to 1 for every attempt because it's the first attempt for a user on a given exam?

Reply:
Also should we be setting this value on attempt creation? The first example is giving me pause.

Reply:
That's what it sounds like to me. It's the combination of the `resource_link_id` (some unique identifier for the exam - presumably the `resource_id` model field on the exam), `sub`, and `attempt_number` claims that uniquely identify an attempt in the LTI launch.

We could set up a composite primary key on (user, exam, attempt_number).

Reply:
Because we don't allow multiple attempts for the same exam right now, wouldn't that mean the attempt_code on creation should always be 1?

I know multiple attempts is out of scope right now, but just trying to think about how to add the right functionality/make it easy to add multiple attempts in the future